### PR TITLE
DEV9: Always bind UDP ports in sockets

### PR DESCRIPTION
**Pull request by TheLastRar**
### Description of Changes
Always bind the UDP source port when using sockets
Reworded some logs
Duplicate FixedPort's connections vector for resetting child connections

### Rationale behind Changes
Previously, we would only bind the UDP source port in specific situations (Broadcast, multicast and when src and dst port similar).
This proved inadequate for some games, which would either fail to connect of behave strangely if the source port wasn't maintained.
Expanding the heuristics used to catch every game seemed non-trivial, so I've opted to forgo this previously used heuristic and always bind the port.

This does mean we now sometimes bind the port unnecessarily (namely DNS requests), however, different games use ranges for dynamic ports, so avoiding this seemed not worth the effort.

Additionally, FixedPort's `Reset()` method could potentially iterate over a modified vector if the child connection closed itself (and thus got removed from the connections vector in FixedPort's connection closed handler).
Duplicating this vector avoids that situation.
This also allows us to reduce the scope of the lock in `Reset()`, avoiding a deadlock (or crash) as FixedPort's connection closed handler would also try to lock.

### Suggested Testing Steps
Test networking using Sockets

### CODE REVIEW

**Comment by AmyRoxwell** [comment]
Would be testing tonight.

**Review by AmyRoxwell — APPROVED** [review]
MH2DOS Online still works on linux, even on really heavy situations.
Monster Hunter 2_SLPM-66280_20241128225107
Emulog for reference: i forgot to sent it oops
[emulog.zip](https://github.com/user-attachments/files/17954917/emulog.zip)

File: pcsx2/DEV9/Sessions/UDP_Session/UDP_FixedPort.cpp [hunk 1]
```diff
@@ -225,10 +225,18 @@ namespace Sessions
 
 	void UDP_FixedPort::Reset()
 	{
-		std::lock_guard numberlock(connectionSentry);
+		// Reseting a session may cause that session to close itself,
+		// hen that happens, connections gets modified via our close handler.
```
Comment (F0bes) [inline pcsx2/DEV9/Sessions/UDP_Session/UDP_FixedPort.cpp:229]: typo

Comment (TheLastRar) [reply pcsx2/DEV9/Sessions/UDP_Session/UDP_FixedPort.cpp]: Amended, Also slightly reworded later text

**Review by Mrlinkwii — APPROVED** [review]

**Comment by IamLupo** [comment]
This update causes issues on the games:
Battlefield 2: Modern combat and Medal of honor rising sun.

For Battlefield 2: Modern combat. If we enable punkbuster on the game server we let the client connect to "bfmc1.evenbalance.com". We have pointed this domain to our backend services. Then he tries to connect to this services through UDP port 9867. Because haven't recreate this services our backend wont return anything.
This update causes that the emulator freezes and don't respond anymore because it expects a UDP response. Instead it gets a ICMP "Destination unreachable". I only can close pcsx2 with task manager.

For Medal of Honor - Rising Sun we have the following case. When we go online it resolves the domain name "mohrs.theater.ea.com" to our backend services. When we first go online it uses 14300 TCP to use the EA Theater protocol. We have developed all these protocol calls and resolved everything.
If you want to create a game it start to use 28500 UDP to our backend services with an "ECHO" command. We never had implemented this call. So we never gave a response to this request.
So basically the same thing happpends. The emulator freezes and don't respond anymore because it expects a UDP response. Instead it gets a ICMP "Destination unreachable". I only can close pcsx2 with task manager.

As you can understand is that earlier pcsx2 versions never froze the hole emulator when a UDP reponse was never send. Like I expect some thread is stuck and make the emulator freeze.

I narrowed down this version by testing:
```
pcsx2-v2.3.40-windows-x64-Qt Not Effected
pcsx2-v2.3.41-windows-x64-Qt Effected
pcsx2-v2.3.42-windows-x64-Qt Effected
pcsx2-v2.3.43-windows-x64-Qt Effected
pcsx2-v2.3.44-windows-x64-Qt Effected
pcsx2-v2.3.45-windows-x64-Qt Effected
```

How can i solve this issue?
For Battlefield 2: Modern combat we disable punkbuster and the ps2 game never connects to "bfmc1.evenbalance.com" anymore. So no UDP connection will be opened.
For Medal of Honor we should still implement a response for the ECHO command on UDP 28500. Probably this should resolve this issue.

I hope you guys can give some intuition why this update makes the emulator freeze and not respond anymore.

**Comment by TheLastRar** [comment]
Can you provide logs for the failures
Additionally what OS are you on?

**Comment by IamLupo** [comment]
> Can you provide logs for the failures Additionally what OS are you on?

OS: Windows
[emulog.txt](https://github.com/user-attachments/files/19687801/emulog.txt)

**Comment by IamLupo** [comment]
These are the logs from Medal of Honor - Rising Sun
[emulog.txt](https://github.com/user-attachments/files/19687926/emulog.txt)

**Comment by IamLupo** [comment]
Here you see a video maybe that will help: https://www.youtube.com/watch?v=S5Qknza7Xa4

**Comment by IamLupo** [comment]
This is the emulogs for the pcsx2-v2.3.40-windows-x64-Qt  that is not affected.
[emulog.txt](https://github.com/user-attachments/files/19706984/emulog.txt)

2.3.40:

  [   51.6283] Setting vsync mode: Disabled
  [   52.1314] DEV9: Socket: Creating New UDP Connection to 14300
  [   52.2420] DEV9: UDP: Recv error: 10054
  [   52.2420] DEV9: Socket: Closed Dead UDP Connection to 14300
  [   58.5023] DEV9: Socket: Creating New UDP Connection to 14300
  [   58.6120] DEV9: UDP: Recv error: 10054
  [   58.6120] DEV9: Socket: Closed Dead UDP Connection to 14300

2.3.41:

  [   51.1205] Setting vsync mode: Disabled
  [   53.1086] DEV9: Socket: Binding UDP fixed port 28500
  [   53.1087] DEV9: Socket: Creating New UDP Connection from fixed port 28500 to 14300
  [   53.2225] DEV9: UDP: UDP recv error: 10054
  [   53.2225] DEV9: Socket: Unbound fixed port 28500
  [   59.3789] DEV9: UDP: Send error 10038

Looking at this makes me really confused,.. why is it in 2.3.40 UDP 14300 and in version 2.3.41 UDP 28500?

**Comment by IamLupo** [comment]
Oke i notice that it always make it a UDP_FixedPort class. The check if source and destination < 10 got removed in this update to focus it to make a normal UDP_Session.

I changed the code like this to fix it:

  bool SocketAdapter::SendUDP(ConnectionKey Key, IP_Packet* ipPkt)
  {
	  IP_PayloadPtr* ipPayload = static_cast<IP_PayloadPtr*>(ipPkt->GetPayload());
	  UDP_Packet udp(ipPayload->data, ipPayload->GetLength());
  
	  Key.ps2Port = udp.sourcePort;
	  Key.srvPort = udp.destinationPort;
  
	  const int res = SendFromConnection(Key, ipPkt);
	  if (res == 1)
		  return true;
	  else if (res == 0)
		  return false;
	  else
	  {
		  UDP_Session* s = nullptr;
  
		  if (abs(udp.sourcePort - udp.destinationPort) <= 10 || //Used for games that assume the destination/source port
			  ipPkt->destinationIP == dhcpServer.broadcastIP || //Broadcast packets
			  ipPkt->destinationIP == IP_Address{{{255, 255, 255, 255}}} || //Limited Broadcast packets
			  (ipPkt->destinationIP.bytes[0] & 0xF0) == 0xE0) //Multicast address start with 0b1110
		  {
			  // Always bind the UDP source port
			  // PS2 software can run into issues if the source port is not preserved
			  UDP_FixedPort* fPort = nullptr;
			  BaseSession* fSession;
			  if (connections.TryGetValue(Key, &fSession))
			  {
				  fPort = static_cast<UDP_FixedPort*>(fSession);
			  }
			  else
			  {
				  ConnectionKey fKey{};
				  fKey.protocol = static_cast<u8>(IP_Type::UDP);
				  fKey.ps2Port = udp.sourcePort;
				  fKey.srvPort = 0;
  
				  Console.WriteLn("DEV9: Socket: Binding UDP fixed port %d", udp.sourcePort);
  
				  fPort = new UDP_FixedPort(fKey, adapterIP, udp.sourcePort);
				  fPort->AddConnectionClosedHandler([&](BaseSession* session) { HandleFixedPortClosed(session); });
  
				  fPort->destIP = {};
				  fPort->sourceIP = dhcpServer.ps2IP;
  
				  connections.Add(fKey, fPort);
				  fixedUDPPorts.Add(udp.sourcePort, fPort);
  
				  fPort->Init();
			  }
  
			  Console.WriteLn("DEV9: Socket: Creating New UDP Connection from fixed port %d to %d", udp.sourcePort, udp.destinationPort);
			  s = fPort->NewClientSession(Key,
				  ipPkt->destinationIP == dhcpServer.broadcastIP || ipPkt->destinationIP == IP_Address{{{255, 255, 255, 255}}},
				  (ipPkt->destinationIP.bytes[0] & 0xF0) == 0xE0);
  
			  // If we are unable to bind to the port, fall back to a dynamic port
			  if (s == nullptr)
			  {
				  Console.Error("DEV9: Socket: Failed to Create New UDP Connection from fixed port");
				  Console.WriteLn("DEV9: Socket: Retrying with dynamic port to %d", udp.destinationPort);
				  s = new UDP_Session(Key, adapterIP);
			  }
		  }
		  else
		  {
			  Console.WriteLn("DEV9: Socket: Creating New UDP Connection to %d", udp.destinationPort);
			  s = new UDP_Session(Key, adapterIP);
		  }
  
		  s->AddConnectionClosedHandler([&](BaseSession* session) { HandleConnectionClosed(session); });
		  s->destIP = ipPkt->destinationIP;
		  s->sourceIP = dhcpServer.ps2IP;
		  connections.Add(Key, s);
		  return s->Send(ipPkt->GetPayload());                                           
	  }
  }

**Comment by TheLastRar** [comment]
> This is the emulogs for the pcsx2-v2.3.40-windows-x64-Qt that is not affected. [emulog.txt](https://github.com/user-attachments/files/19706984/emulog.txt)
> 
> Looking at the data i get:
> 

The difference in the logs between is that fixed port logs both the source and destination ports
`Binding UDP fixed port 28500` logs the source port, not the destination port

> Keep in mind i totally ignore the message:
> 
> ```c++
>     // Always bind the UDP source port
>     // PS2 software can run into issues if the source port is not preserved
> ```
> 
> Like i dont know what kind of issues this causes.

Anything that performs UDP hole punching needs the source port preserved
If the source port is ignored (allowing the OS / normal UDP_Session to remap the port)
SOCOM's voice chat will fail to work.
Syphon Filter: The Omega Strain not connect _correctly_, to a game lobby, leading to a bugged player.
LAN also needs the source port preserved, but other issues in Sockets prevent LAN.

Are you able to test this PR https://github.com/PCSX2/pcsx2/pull/12516 to see if it fixes you issue

**Comment by IamLupo** [comment]
I am in bed right now, I will test it tomorrow!

**Comment by IamLupo** [comment]
> > This is the emulogs for the pcsx2-v2.3.40-windows-x64-Qt that is not affected. [emulog.txt](https://github.com/user-attachments/files/19706984/emulog.txt)
> > Looking at the data i get:
> 
> The difference in the logs between is that fixed port logs both the source and destination ports (and v2.3.41 used Fixed ports for all UDP connections) `Binding UDP fixed port 28500` logs the source port, not the destination port
> 
> > Keep in mind i totally ignore the message:
> > ```c++
> >     // Always bind the UDP source port
> >     // PS2 software can run into issues if the source port is not preserved
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Like i dont know what kind of issues this causes.
> 
> Anything that performs UDP hole punching needs the source port preserved If the source port is ignored (allowing the OS / normal UDP_Session to remap the port) SOCOM's voice chat will fail to work. Syphon Filter: The Omega Strain not connect _correctly_, to a game lobby, leading to a bugged player. LAN also needs the source port preserved, but other issues in Sockets prevent LAN.
> 
> Are you able to test this PR #12516 to see if it fixes you issue?

That worked perfectly! No more freezes!

Here are the logs:
```
[   88.9033] Setting vsync mode: Disabled
[   90.7403] DEV9: Socket: Binding UDP fixed port 28500
[   90.7404] DEV9: Socket: Creating New UDP Connection from fixed port 28500 to 14300
[   90.8485] DEV9: UDP: UDP recv error: 10054
[   90.8485] DEV9: Socket: Unbound fixed port 28500
[   90.8485] DEV9: Socket: Closed Dead UDP Connection to 14300
[   95.6436] DEV9: Socket: Binding UDP fixed port 28500
[   95.6436] DEV9: Socket: Creating New UDP Connection from fixed port 28500 to 14300
[   95.7515] DEV9: UDP: UDP recv error: 10054
[   95.7515] DEV9: Socket: Unbound fixed port 28500
[   95.7515] DEV9: Socket: Closed Dead UDP Connection to 14300
[   99.1137] DEV9: Socket: Binding UDP fixed port 28500
[   99.1138] DEV9: Socket: Creating New UDP Connection from fixed port 28500 to 14300
[   99.2207] DEV9: UDP: UDP recv error: 10054
[   99.2208] DEV9: Socket: Unbound fixed port 28500
[   99.2208] DEV9: Socket: Closed Dead UDP Connection to 14300
```

**Comment by TheLastRar** [comment]
> > Are you able to test this PR #12516 to see if it fixes you issue?
> 
> That worked perfectly! No more freezes!
> 
> Here are the logs:
> 
> ```
> [   88.9033] Setting vsync mode: Disabled
> [   90.7403] DEV9: Socket: Binding UDP fixed port 28500
> [   90.7404] DEV9: Socket: Creating New UDP Connection from fixed port 28500 to 14300
> [   90.8485] DEV9: UDP: UDP recv error: 10054
> [   90.8485] DEV9: Socket: Unbound fixed port 28500
> [   90.8485] DEV9: Socket: Closed Dead UDP Connection to 14300
> [   95.6436] DEV9: Socket: Binding UDP fixed port 28500
> [   95.6436] DEV9: Socket: Creating New UDP Connection from fixed port 28500 to 14300
> [   95.7515] DEV9: UDP: UDP recv error: 10054
> [   95.7515] DEV9: Socket: Unbound fixed port 28500
> [   95.7515] DEV9: Socket: Closed Dead UDP Connection to 14300
> [   99.1137] DEV9: Socket: Binding UDP fixed port 28500
> [   99.1138] DEV9: Socket: Creating New UDP Connection from fixed port 28500 to 14300
> [   99.2207] DEV9: UDP: UDP recv error: 10054
> [   99.2208] DEV9: Socket: Unbound fixed port 28500
> [   99.2208] DEV9: Socket: Closed Dead UDP Connection to 14300
> ```

Thanks, looking at the logs there is still a smaller issue remaining
I've pushed an update to the PR, can you grab an updated build and post the logs

**Comment by IamLupo** [comment]
> > > Are you able to test this PR #12516 to see if it fixes you issue?
> > 
> > 
> > That worked perfectly! No more freezes!
> > Here are the logs:
> > ```
> > [   88.9033] Setting vsync mode: Disabled
> > [   90.7403] DEV9: Socket: Binding UDP fixed port 28500
> > [   90.7404] DEV9: Socket: Creating New UDP Connection from fixed port 28500 to 14300
> > [   90.8485] DEV9: UDP: UDP recv error: 10054
> > [   90.8485] DEV9: Socket: Unbound fixed port 28500
> > [   90.8485] DEV9: Socket: Closed Dead UDP Connection to 14300
> > [   95.6436] DEV9: Socket: Binding UDP fixed port 28500
> > [   95.6436] DEV9: Socket: Creating New UDP Connection from fixed port 28500 to 14300
> > [   95.7515] DEV9: UDP: UDP recv error: 10054
> > [   95.7515] DEV9: Socket: Unbound fixed port 28500
> > [   95.7515] DEV9: Socket: Closed Dead UDP Connection to 14300
> > [   99.1137] DEV9: Socket: Binding UDP fixed port 28500
> > [   99.1138] DEV9: Socket: Creating New UDP Connection from fixed port 28500 to 14300
> > [   99.2207] DEV9: UDP: UDP recv error: 10054
> > [   99.2208] DEV9: Socket: Unbound fixed port 28500
> > [   99.2208] DEV9: Socket: Closed Dead UDP Connection to 14300
> > ```
> 
> Thanks, looking at the logs there is still a smaller issue remaining I've pushed an update to the PR, can you grab an updated build and post the logs

Still works. Logs looks a bit different because it now only makes once the bind. Also no udp disconnect message anymore. In wireshark all looks good 👍 

```
[   59.7921] Setting vsync mode: Disabled
[   62.1634] DEV9: Socket: Binding UDP fixed port 28500
[   62.1634] DEV9: Socket: Creating New UDP Connection from fixed port 28500 to 14300
[   62.2711] DEV9: UDP: recvfrom error: 10054
[   68.9431] DEV9: UDP: recvfrom error: 10054
[   88.4615] DEV9: UDP: recvfrom error: 10054
```

[emulog.txt](https://github.com/user-attachments/files/19726279/emulog.txt)

**Comment by TheLastRar** [comment]
> Still works. Logs looks a bit different because it now only makes once the bind. Also no udp disconnect message anymore. In wireshark all looks good 👍
> 
> ```
> [   59.7921] Setting vsync mode: Disabled
> [   62.1634] DEV9: Socket: Binding UDP fixed port 28500
> [   62.1634] DEV9: Socket: Creating New UDP Connection from fixed port 28500 to 14300
> [   62.2711] DEV9: UDP: recvfrom error: 10054
> [   68.9431] DEV9: UDP: recvfrom error: 10054
> [   88.4615] DEV9: UDP: recvfrom error: 10054
> ```
> 
> [emulog.txt](https://github.com/user-attachments/files/19726279/emulog.txt)

Yep, that looks as expected, thanks for testing

**Pull request merged**